src/analysis/normed_space/M_structure.lean

```diff
@@ -0,0 +1,380 @@
+$$
+∥x∥ = ∥P x∥ + ∥(1-P) x∥.
+$$
+-/
+def is_Lprojection (P : X →L[𝕜] X) : Prop := is_projection P ∧ ∀ (x : X), ∥x∥ = ∥P x∥ + ∥(1-P) x∥
+
+/--
+A projection on a normed space `X` is said to be an M-projection if, for all `x` in `X`,
+$$
+∥x∥ = max(∥P x∥, ∥(1-P) x∥).
+$$
+-/
+def is_Mprojection (P: X →L[𝕜] X) : Prop :=
+  is_projection P ∧ ∀ (x : X), ∥x∥ = (max ∥P x∥  ∥(1-P) x∥)
+
+namespace is_Lprojection
+
+lemma Lcomplement {P: X →L[𝕜] X} (h: is_Lprojection P) :  is_Lprojection (1-P) :=
+⟨is_projection.complement_iff.mp h.1, λ x, by { rw [add_comm, sub_sub_cancel], exact h.2 x }⟩
+
+lemma Lcomplement_iff (P: X →L[𝕜] X) : is_Lprojection P ↔ is_Lprojection (1-P) :=
+⟨Lcomplement, λ h, by { rw [← sub_sub_cancel 1 P], exact Lcomplement h }⟩
+
+lemma PQ_eq_QPQ (P Q : X →L[𝕜] X) (h₁: is_Lprojection P) (h₂: is_Lprojection Q) :
+  P * Q = Q * P * Q :=
+begin
+  ext,
+  rw ← norm_sub_eq_zero_iff,
+  have e1 : ∥Q x∥ ≥ ∥Q x∥ + 2 • ∥ (P * Q) x - (Q * P * Q) x∥ :=
+  calc ∥Q x∥ = ∥P (Q x)∥ + ∥(1 - P) (Q x)∥ : by rw h₁.right
+  ... = ∥Q (P (Q x))∥ + ∥(1-Q) (P (Q x))∥ + ∥(1 - P) (Q x)∥ : by rw h₂.right
+  ... = ∥Q (P (Q x))∥ + ∥(1-Q) (P (Q x))∥ + (∥Q ((1 - P) (Q x))∥ + ∥(1-Q) ((1 - P) (Q x))∥) :
+    by rw h₂.right ((1 - P) (Q x))
+  ... = ∥Q (P (Q x))∥ + ∥(1-Q) (P (Q x))∥ + (∥Q (Q x - P (Q x))∥ + ∥(1-Q) ((1 - P) (Q x))∥) : rfl
+  ... = ∥Q (P (Q x))∥ + ∥(1-Q) (P (Q x))∥ + (∥Q (Q x) - Q (P (Q x))∥ + ∥(1-Q) ((1 - P) (Q x))∥) :
+    by rw map_sub
+  ... = ∥Q (P (Q x))∥ + ∥(1-Q) (P (Q x))∥ + (∥(Q * Q) x - Q (P (Q x))∥ + ∥(1-Q) ((1 - P) (Q x))∥) :
+    rfl
+  ... = ∥Q (P (Q x))∥ + ∥(1-Q) (P (Q x))∥ + (∥Q x - Q (P (Q x))∥ + ∥(1-Q) ((1 - P) (Q x))∥) :
+    by rw [← sq, projection_def h₂.left]
+  ... = ∥Q (P (Q x))∥ + ∥(1-Q) (P (Q x))∥ + (∥Q x - Q (P (Q x))∥ + ∥(1-Q) (Q x - P (Q x))∥) : rfl
+  ... = ∥Q (P (Q x))∥ + ∥(1-Q) (P (Q x))∥ + (∥Q x - Q (P (Q x))∥ + ∥(1-Q) (Q x) - (1-Q) (P (Q x))∥)
+    : by rw map_sub
+  ... = ∥Q (P (Q x))∥ + ∥(1-Q) (P (Q x))∥ + (∥Q x - Q (P (Q x))∥
+    + ∥((1-Q) * Q) x - (1-Q) (P (Q x))∥) : rfl
+  ... = ∥Q (P (Q x))∥ + ∥(1-Q) (P (Q x))∥ + (∥Q x - Q (P (Q x))∥ + ∥0 - (1-Q) (P (Q x))∥) :
+    by {rw [sub_mul, ← sq, projection_def h₂.left, one_mul, sub_self ], exact rfl }
+  ... = ∥Q (P (Q x))∥ + ∥(1-Q) (P (Q x))∥ + (∥Q x - Q (P (Q x))∥ + ∥(1-Q) (P (Q x))∥) :
+    by rw [zero_sub, norm_neg]
+  ... = ∥Q (P (Q x))∥ + ∥Q x - Q (P (Q x))∥ + 2•∥(1-Q) (P (Q x))∥  : by abel
+  ... ≥ ∥Q x∥ + 2 • ∥ (P * Q) x - (Q * P * Q) x∥ :
+    by exact add_le_add_right (norm_le_insert' (Q x) (Q (P (Q x)))) (2•∥(1-Q) (P (Q x))∥),
+  rw ge at e1,
+  nth_rewrite_rhs 0 ← add_zero (∥Q x∥) at e1,
+  rw [add_le_add_iff_left, two_smul,  ← two_mul]  at e1,
+  rw le_antisymm_iff,
+  refine ⟨_, norm_nonneg _⟩,
+  rwa [←mul_zero (2:ℝ), mul_le_mul_left (show (0:ℝ) < 2, by norm_num)] at e1
+end
+
+lemma QP_eq_QPQ (P Q : X →L[𝕜] X) (h₁: is_Lprojection P) (h₂: is_Lprojection Q) : Q * P = Q * P * Q
```

Comment:
Also the lemma should be named in a way that doesn't rely on the names of it's arguments

Reply:
> It seems like the right- and left-hand sides should be switched to me.

@j-loreaux  Is that so that the simpler side is on the right (as for simp lemmas)?

Reply:
for names, after the switching of left right, perhaps this should be `mul_mul_self_eq_self_mul`? and the other would be `mul_mul_self_eq_mul_self`?

Reply:
Now that the proofs are a bit shorter I was wondering about moving `PQ_eq_QPQ` and `QP_eq_QPQ` to `have` statements under the proof of `Lproj_commute` since they're just intermediary steps to proving that.

Reply:
If I make `PQ_eq_QPQ` and `QP_eq_QPQ` to `have` statements is it still necessary to put the simpler side on the right?

Reply:
No.

Reply:
I've renamed `Lproj_commute` as `commute` and I've moved `PQ_eq_QPQ` and `QP_eq_QPQ` into the proof of is_Lprojection.commute which I think resolves this point.